### PR TITLE
Examples can now be run without errors, also some typos fixed

### PR DESCRIPTION
**Pull request by matthijsdewit111**
SolverBase.task was never set, so removed
SolverOrbital.chkpt_every was not always set, so use get_attr with default value for logging attributes
SolverBase.save_checkpoint() no longer returns loss
plot_data couldn't find attribute, now plots attribute of best model

### CODE REVIEW

File: qmctorch/solver/solver_base.py [hunk 1]
```diff
@@ -30,10 +30,9 @@ def __init__(self, wf=None, sampler=None,
         self.scheduler = scheduler
         self.cuda = False
         self.device = torch.device('cpu')
-        self.task = None
 
         # if pos are needed for the optimizer (obsolete ?)
-        if 'lpos_needed' not in self.opt.__dict__.keys():
+        if self.opt is not None and 'lpos_needed' not in self.opt.__dict__.keys():
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_base.py:35]: good catch ! We can potentially not have an optimizer !

File: qmctorch/solver/solver_orbital.py [hunk 2]
```diff
@@ -159,7 +159,7 @@ def restore_sampling_parameters(self):
 
     def geo_opt(self, nepoch, geo_lr=1e-2, batchsize=None,
                 nepoch_wf_init=100, nepoch_wf_update=50,
-                hdf5_group=None, chkpt_every=None, tqdm=False):
+                hdf5_group='geo_opt', chkpt_every=None, tqdm=False):
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_orbital.py:162]: great thanks !

File: qmctorch/solver/solver_base.py [hunk 3]
```diff
@@ -116,9 +115,6 @@ def track_observable(self, obs_name):
         if 'energy' not in obs_name:
             obs_name += ['energy']
 
-        if self.task == 'geo_opt' and 'geometry' not in obs_name:
-            obs_name += ['geometry']
-
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_base.py:121]: why should we remove that part ?

File: qmctorch/solver/solver_orbital.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def geo_opt(self, nepoch, geo_lr=1e-2, batchsize=None,
 
         # log data
         self.prepare_optimization(batchsize, None, tqdm)
-        self.log_data_opt(nepoch, 'geometry optimization')
+        self.log_data_opt(nepoch, hdf5_group)
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_orbital.py:189]: oh yeah that's better :)

File: qmctorch/solver/solver_orbital.py [hunk 5]
```diff
@@ -241,14 +241,14 @@ def run(self, nepoch, batchsize=None,
...
-                                        Defaults to wf.task.
+                                        Defaults to 'wf_opt'.
             chkpt_every (int, optional): save a checkpoint every every iteration.
                                          Defaults to half the number of epoch
         """
 
         # prepare the optimization
         self.prepare_optimization(batchsize, chkpt_every, tqdm)
-        self.log_data_opt(nepoch, 'wave function optimization')
+        self.log_data_opt(nepoch, hdf5_group)
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_orbital.py:251]: I'm not sure what's best here. The group in the hdf5 file could be different from the task. But I agree that it could be simplified.

File: qmctorch/solver/solver_orbital.py [hunk 6]
```diff
@@ -459,5 +459,5 @@ def log_data_opt(self, nepoch, task):
         log.info(
             '  Output file         : {0}', self.hdf5file)
         log.info(
-            '  Checkpoint every    : {0}', self.chkpt_every)
+            '  Checkpoint every    : {0}', getattr(self, 'chkpt_every', None))
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_orbital.py:462]: why is that needed ?

File: qmctorch/solver/solver_orbital_horovod.py [hunk 7]
```diff
@@ -47,23 +47,23 @@ def __init__(self, wf=None, sampler=None, optimizer=None,
         self.sampler.walkers.nwalkers //= hvd.size()
 
     def run(self, nepoch, batchsize=None, loss='energy',
-            clip_loss=False, grad='manual', hdf5_group=None,
-            num_threads=1):
+            clip_loss=False, grad='manual', hdf5_group='optimization',
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_orbital_horovod.py:50]: in principle we could do wave function or geometry optimization. So maybe de default group could be wf_opt

**Review by NicoRenaud — CHANGES_REQUESTED** [review]
Thanks good work ! I have very minor comment in the PR.
One thing though is that the solver_horovod is never tested on GitHub Action. However we can run MPI processes on GA so i think we should be able to have a very small test just to make sure that the solver runs. Could you look into that since you fixed the solver_horovod in that PR ?

File: qmctorch/solver/solver_base.py (hunk 3 above)
Comment (matthijsdewit111) [reply qmctorch/solver/solver_base.py:121]: We can keep the second part of the if statement, I'll put it back in. The first part of the if statement isn't needed since self.task is never assigned I found.

File: qmctorch/solver/solver_orbital.py (hunk 6 above)
Comment (matthijsdewit111) [reply qmctorch/solver/solver_orbital.py:462]: When I ran the Horovod example, it never executed SolverOrbital.prepare_optimization() (where self.chkpt_every is assigned), so we should either assign it always somehow or make it optional, I agree this right now is not the best solution, but I just wanted it to run :)
I think it has to do with the fact that SolverOrbitalHorovod never used the chkpt_every before, it only saved checkpoints if it found a lower energy state.
So should I change SolverOrbitalHorovod.run() to saving a checkpoint every few iteration and making sure it sets self.chkpt_every, or we make it optional and just default to None in the SolverOrbital.init(), so the attribute always exists at least.

**Comment by matthijsdewit111** [comment]
Still working on testing horovod with github actions.
(I spend some time trying to run github actions locally, but after spending much time, I'm going to try by just committing changes to git I think and see if I can get it to work)

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -38,3 +39,9 @@ jobs:
         CONDA_PREFIX: /usr/share/miniconda
       run: |
         pytest --cov-config setup.cfg
+
+    - name: Test with multithreading
+      env:
+        CONDA_PREFIX: /usr/share/miniconda
+      run: |
+        pytest horovod_tests -n 2
```
Comment (NicoRenaud) [inline .github/workflows/build.yml:47]: Great that it works !! 
can  you have coverage as well by adding the --cov-config or something like that ?

File: horovod_tests/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+__all__ = ["PATH_QMCTORCH", "PATH_TEST", "second_derivative"]
+
+# Environment data
+PATH_QMCTORCH = Path(pkg.resource_filename('qmctorch', ''))
+ROOT = PATH_QMCTORCH.parent
+
+PATH_TEST = ROOT / "horovod_tests"
+
+
+def second_derivative(xm1, x0, xp1, eps):
```
Comment (NicoRenaud) [inline horovod_tests/test_utils.py:15]: I'm not sure the second_derivative is really needed ... my bad

File: qmctorch/solver/solver_orbital_horovod.py (hunk 7 above)
Comment (NicoRenaud) [reply qmctorch/solver/solver_orbital_horovod.py:50]: I will still change the default name to wf_opt :)

File: qmctorch/solver/solver_orbital_horovod.py [hunk 10]
```diff
@@ -171,11 +171,15 @@ def run(self, nepoch, batchsize=None, loss='energy',
...
-                    min_loss = self.save_checkpoint(
-                        n, cumulative_loss, self.save_model)
+                if n == 0 or cumulative_loss < min_loss:
+                    self.observable.models.best = dict(
+                        self.wf.state_dict())
+                min_loss = cumulative_loss
+
+                if self.chkpt_every is not None:
+                    if (n > 0) and (n % chkpt_every == 0):
+                        self.save_checkpoint(n, cumulative_loss)
```
Comment (NicoRenaud) [inline qmctorch/solver/solver_orbital_horovod.py:181]: Nice !

**Review by NicoRenaud — CHANGES_REQUESTED** [review]
It seems to be working ! Nicely done !
Could you see if you can get code coverage as well for thenhvd test ?

**Review by NicoRenaud — APPROVED** [review]
Very nice work ! You can merge when you think you're done  (sorry I was so late for approving the PR)
